scipy/optimize/_numdiff.py

```diff
@@ -570,3 +638,48 @@ def check_derivative(fun, jac, x0, bounds=(-np.inf, np.inf), args=(),
                                    args=args, kwargs=kwargs)
         abs_err = np.abs(J_to_test - J_diff)
         return np.max(abs_err / np.maximum(1, np.abs(J_diff)))
+
+
+class FiniteDifference:
+    """Represents finite difference approximation.
+
+    Provide an easy way of calling `approx_derivative` repeatedly
+    with the same set of parameters. Avoid duplicated calls to ``fun``
+    if ``avoid_duplicated_calls=True``.
+    """
+
+    def __init__(self, fun, method='3-point', rel_step=None,
+                 bounds=(-np.inf, np.inf), sparsity=None,
+                 as_linear_operator=False, x0=None,
+                 avoid_duplicated_calls=True):
+        self._fun = fun
+        self.method = method
+        self.rel_step = rel_step
+        self.bounds = bounds
+        self.sparsity = sparsity
+        self.as_linear_operator = as_linear_operator
+        self.avoid_duplicated_calls = avoid_duplicated_calls
+        self._x = None
+        self._args = ()
+        self._kwargs = {}
+        self._f = None
+
+    def fun_wrapped(self, x, *args, **kwargs):
+        if not np.array_equal(self._x, x) or \
+           not args == self._args or \
```

Comment:
Between the three approaches you have mentioned:
1. Explain limitations on documentation;
2. Fix ``args`` and ``kwargs`` in the constructor;
3. Remove this feature.

I like option (1) better. Mostly because I want to use this function like this:
```
>>> hess = FiniteDifference(jac_dot_v, ... )
>>> hess(x1, v1)
>>> hess(x2, v1)
>>> hess(x3, v1)
>>> hess(x1, v2)
>>> hess(x2, v2)
>>> hess(x3, v2)
...
```
And it would be useful to have this feature when doing this.

Furthermore, while this feature may fail to avoid some duplicated calls, I don't see how it could end up providing the wrong result. In other words, the worst case scenario would be a worse performance.